scene/gui/text_edit.cpp

```diff
@@ -2924,6 +2922,35 @@ void TextEdit::_update_ime_text() {
 	queue_redraw();
 }
 
+void TextEdit::_on_focus_enter() {
+	if (caret_blink_enabled) {
+		caret_blink_timer->start();
+	} else {
+		draw_caret = true;
+	}
+
+	_update_ime_window_position();
+
+	if (virtual_keyboard_enabled && DisplayServer::get_singleton()->has_feature(DisplayServer::FEATURE_VIRTUAL_KEYBOARD)) {
```

Comment:
```suggestion
void TextEdit::_show_virtual_keyboard() {
	if (virtual_keyboard_enabled && DisplayServer::get_singleton()->has_feature(DisplayServer::FEATURE_VIRTUAL_KEYBOARD)) {
```

This can be just about the virtual keyboard, the caret blink stuff and the ime window update can be kept back in the focus enter notification. When the mouse is released, the caret blink doesn't need to be updated and a mouse press closes the ime so it should never be up when a mouse release happens.

Reply:
i was hesitant about changing any of the actual behavior of the text editor, i wanted to keep these as strictly optimizations, im not sure if  `_update_ime_window_position` is ok to remove here? it seems important for showing the virtual keyboard
but ill refactor the caret blink stuff.